File: ruby/Problem021_AmicableNumbers.rb
require "prime"

MAX = 10000

def proper_deviders_sum input
    sum = 1

    (2..Math.sqrt(input)).each do |i|
        if input % i == 0
            sum += i
            sum += input / i
        end
    end

    return sum
end


sum = 0
(1...MAX).each do |i|
    v1 = proper_deviders_sum(i)
    v2 = proper_deviders_sum(v1)

    if (i == v2 && v2 != v1)
        sum += i
    end

end

puts "sum: #{sum}"

File: cpp/Problem021_AmicableNumbers.cpp
#include <iostream>
#include <cmath>
#define MAX 10000
using std::cout;

int d(int a)
{
    int b = 1;
    for (int i=2; i<=int(std::sqrt(a)); i++)
        if (a % i == 0)
            b += i + int(a/i);
    return b;
}

int main()
{
    int cache[MAX] = {};
    int sum = 0;
    int a = 1;
    while (a++ < MAX)
    {
        if (cache[a] != 0)
            continue;
        
        int b = d(a);
        if (a != b and d(b) == a)
        {
            sum += a + b;
            cache[b] = 999;
        }
            
    }

    cout << sum;
    return 0;
}